src/rpc/stats.cpp

```diff
@@ -0,0 +1,203 @@
+            auto historyObj = entry.get_obj();
+            StatHistoryEntry historyEntry {
+                historyObj["timestamp"].get_int64(),
+                historyObj["latency"].get_int64(),
+                historyObj["payload"].get_int64()
+            };
+            stats.history.push_back(historyEntry);
+        }
+    }
+    return stats;
+}
+
+void CRPCStats::add(const std::string& name, const int64_t latency, const int64_t payload)
+{
+    auto stats = CRPCStats::get(name);
+    if (stats) {
+        stats->count++;
+        stats->lastUsedTime = GetSystemTimeInSeconds();
+        stats->latency = {
+            std::min(latency, stats->latency.min),
+            stats->latency.avg + (latency - stats->latency.avg) / stats->count,
+            std::max(latency, stats->latency.max)
+        };
+        stats->payload = {
+            std::min(payload, stats->payload.min),
+            stats->payload.avg + (payload - stats->payload.avg) / stats->count,
+            std::max(payload, stats->payload.max)
+        };
+    } else {
+        stats = { name, latency, payload };
+    }
+    stats->history.push_back({ stats->lastUsedTime, latency, payload });
+
+    CLockFreeGuard lock(lock_stats);
+    map[name] = *stats;
+}
+
+UniValue CRPCStats::toJSON() {
+    CLockFreeGuard lock(lock_stats);
+
+    UniValue ret(UniValue::VARR);
+    for (auto &[_, stats] : map) {
+        ret.push_back(stats.toJSON());
+    }
+    return ret;
+}
+
+static UniValue getrpcstats(const JSONRPCRequest& request)
+{
+    RPCHelpMan{"getrpcstats",
+        "\nGet RPC stats for selected command.\n",
+        {
+            {"command", RPCArg::Type::STR, RPCArg::Optional::NO, "The command to get stats for."}
+        },
+        RPCResult{
+            " {\n"
+            "  \"name\":               (string) The RPC command name.\n"
+            "  \"latency\":            (json object) Min, max and average latency.\n"
+            "  \"payload\":            (json object) Min, max and average payload size in bytes.\n"
+            "  \"count\":              (numeric) The number of times this command as been used.\n"
+            "  \"lastUsedTime\":       (numeric) Last used time as timestamp.\n"
+            "  \"history\":            (json array) History of last 5 RPC calls.\n"
+            "  [\n"
+            "       {\n"
+            "           \"timestamp\": (numeric)\n"
+            "           \"latency\":   (numeric)\n"
+            "           \"payload\":   (numeric)\n"
+            "       }\n"
+            "  ]\n"
+            "}"
+        },
+        RPCExamples{
+            HelpExampleCli("getrpcstats", "getblockcount") +
+            HelpExampleRpc("getrpcstats", "\"getblockcount\"")
+        },
+    }.Check(request);
+
+    if (!statsRPC.isActive()) {
+        throw JSONRPCError(RPC_INVALID_REQUEST, "Rpcstats is desactivated.");
+    }
+
+    auto command = request.params[0].get_str();
+    auto stats = statsRPC.get(command);
```

Comment:
Example of another method that keeps the reference. (see above).

Reply:
Nope, same as above.

Reply:
Agreed. Same as above.

Reply:
Agreed. Same as above.